src/ipred.rs

```diff
@@ -781,37 +781,36 @@ fn get_upsample(wh: c_int, angle: c_int, is_sm: bool) -> bool {
 }
 
 #[inline(never)]
-unsafe fn upsample_edge<BD: BitDepth>(
-    out: *mut BD::Pixel,
+fn upsample_edge<BD: BitDepth>(
+    out: &mut [BD::Pixel],
     hsz: c_int,
-    in_0: *const BD::Pixel,
+    r#in: &[BD::Pixel; SCRATCH_EDGE_LEN],
+    in_off: usize,
     from: c_int,
     to: c_int,
     bd: BD,
 ) {
     static kernel: [i8; 4] = [-1, 9, 9, -1];
-    let mut i;
-    i = 0 as c_int;
-    while i < hsz - 1 {
-        *out.offset((i * 2) as isize) = *in_0.offset(iclip(i, from, to - 1) as isize);
+    for i in 0..hsz - 1 {
+        out[(i * 2) as usize] = r#in[in_off + iclip(i, from, to - 1) as usize];
```

Comment:
If `from` and `to - 1` are both `usize`s, can we cast to `usize` for them separately, outside of the loop?

Reply:
`from` can't be a `usize` because it's -1 for one of the calls, and so it doesn't make sense to cast `to` to a `usize` because we need `from` and `to` to be the same type for `clip`.